### PR TITLE
sys/checksum: Adding three new crc16 variations

### PR DESCRIPTION
**Pull request by Teufelchen1**
Moin! 👋 

### Contribution description

This PR adds three new crc16 variations: 
* ccitt-false
* ccitt-kermit
* mcrf4xx

The old "ccitt" was renamed to `ccitt-aug` as this is the "correct name" for the given implementation.
Since CRCs aren't really standardised, the names are vague, many have alias. I tried to stick close to what other software projects use (linux, pypi.org/project/crccheck, etc.). 

### Testing procedure

I kept the original test - duplicated & renamed it accordingly.

### CODE REVIEW

**Comment by miri64** [comment]
> The old "ccitt" was renamed to `ccitt-aug` as this is the "correct name" for the given implementation.

This might lead to confusion for users that use the old `ccitt`. Maybe, it is best to move this through proper API aliasing and deprecation, i.e. introduce new `crc16_ccitt_*()` inline functions that wrap around `crc16_ccitt_aug_*` and add a `@deprecation` note that this function will be removed in 2 release cycles (i.e. 2023.01 if this PR gets merged soon). That node could also include a mention that `crc16_ccitt_aug_*` should be used instead.

File: sys/include/checksum/crc16_ccitt.h [hunk 1]
```diff
@@ -39,6 +36,8 @@ extern "C" {
 
 /**
  * @brief           Update CRC16-CCITT
+ * @deprecated      Use @ref crc16_ccitt_false_update instead. 
+ * Will be removed after 2023.01 release.
```
Comment (Teufelchen1) [inline sys/include/checksum/crc16_ccitt.h:40]: > This might lead to confusion for users that use the old ccitt. Maybe, it is best to move this through proper API aliasing and deprecation, i.e. introduce new crc16_ccitt_*() inline functions that wrap around crc16_ccitt_aug_* and add a @deprecation note that this function will be removed in 2 release cycles (i.e. 2023.01 if this PR gets merged soon). That node could also include a mention that crc16_ccitt_aug_* should be used instead.

@miri64 like this? :)

Comment (miri64) [inline sys/include/checksum/crc16_ccitt.h:40]:
```suggestion
 * @deprecated      Use @ref crc16_ccitt_false_update instead.
 *                  Will be removed after 2023.01 release.
```

File: sys/include/checksum/crc16_ccitt.h [hunk 2]
```diff
@@ -53,6 +52,8 @@ uint16_t crc16_ccitt_update(uint16_t crc, const unsigned char *buf, size_t len);
 
 /**
  * @brief           Calculate CRC16-CCITT
+ * @deprecated      Use @ref crc16_ccitt_aug_calc instead. 
+ * Will be removed after 2023.01 release.
```
Comment (miri64) [inline sys/include/checksum/crc16_ccitt.h:56]:
```suggestion
 * @deprecated      Use @ref crc16_ccitt_aug_calc instead.
 *                  Will be removed after 2023.01 release.
```

File: sys/include/checksum/crc16_ccitt.h (hunk 1 above)
Comment (miri64) [reply sys/include/checksum/crc16_ccitt.h:40]: Why is this called `crc16_ccitt_false_update` btw, while the other functions for ccitt-aug are called `crc16_ccitt_aug_...`?

**Review by miri64 — CHANGES_REQUESTED** [review]
> > This might lead to confusion for users that use the old ccitt. Maybe, it is best to move this through proper API aliasing and deprecation, i.e. introduce new crc16_ccitt__() inline functions that wrap around crc16_ccitt_aug__ and add a @Deprecation note that this function will be removed in 2 release cycles (i.e. 2023.01 if this PR gets merged soon). That node could also include a mention that crc16_ccitt_aug_* should be used instead.
> 
> @miri64 like this? :)

Yes, but you also could make this function also now a static inline ... around crc16_ccitt_false_update() now to make the pending deprecation clearer.

Other than that, some formatting issues remain.

File: sys/include/checksum/crc16_ccitt.h (hunk 2 above)
Comment (miri64) [reply sys/include/checksum/crc16_ccitt.h:56]: Same for this one regarding `static inline`...

File: sys/include/checksum/crc16_ccitt.h (hunk 1 above)
Comment (Teufelchen1) [reply sys/include/checksum/crc16_ccitt.h:40]: That's correct. ccitt-false and ccitt-aug only differ in the starting seed. When in doubt, the tests show of this behaviour and how to use it. 
I thought about having a separate function with the suiting name just as a wrapper for naming clarity - but was to shy to do so.

Comment (miri64) [reply sys/include/checksum/crc16_ccitt.h]: > Oh, could I move the wrapper into the header file?

You _should_. Otherwise you don't get much advantage from the inline ;-)

File: sys/include/checksum/crc16_ccitt.h [hunk 3]
```diff
@@ -37,8 +34,92 @@
...
+ * @return          Checksum of the specified memory area based on the
+ *                  given start value
+ */
+uint16_t crc16_ccitt_kermit_update(uint16_t crc, const unsigned char *buf, size_t len);
+
+/**
+ * @brief           Calculate CRC16-CCITT-KERMIT
+ *                  polynom=0x1021
+ *                  seed=0x0000
+ *                  check=0x2189
```
Comment (miri64) [inline sys/include/checksum/crc16_ccitt.h:55]: Be aware, that the `@brief` should only be one line. Maybe it is best to put the polynom, seed, and check into the detail section. E.g.

```suggestion
 *
 * ~~~~~~~~~~~~~~~~~~~~ {.c}
 * polynom=0x1021
 * seed=0x0000
 * check=0x2189
 * ~~~~~~~~~~~~~~~~~~~~ 
```

(this makes it also rendered as code in doxygen)

File: sys/include/checksum/crc16_ccitt.h [hunk 4]
```diff
@@ -37,8 +34,92 @@
...
+ */
+uint16_t crc16_ccitt_false_calc(const unsigned char *buf, size_t len);
+
+/**
+ * @brief           Calculate CRC16-CCITT-AUG
+ *                  polynom=0x1021
+ *                  seed=0x1d0f
+ *                  check=0xe5cc
+ *                  Since CCITT-AUG and CCITT-FALSE only differ in the starting
+ *                  seed, use @ref crc16_ccitt_false_update for updating
```
Comment (miri64) [inline sys/include/checksum/crc16_ccitt.h:110]: IMHO it makes more sense and easier to understand, if there just would be a wrapper:

```C
/**
 * @brief           Update CRC16-CCITT-AUG
 *
 * @param[in]  crc  A start value for the CRC calculation, usually the
 *                  return value of a previous call to
 *                  crc16_ccitt_aug_calc() or crc16_ccitt_aug_update()
 * @param[in]  buf  Start of the memory area to checksum
 * @param[in]  len  Number of bytes to checksum
 *
 * @return          Checksum of the specified memory area based on the
 *                  given start value
 */
static inline uint16_t crc16_ccitt_aug_update(const unsigned char *buf, size_t len)
{
    /* Since CCITT-AUG and CCITT-FALSE only differ in the starting
     * seed, we wrap around crc16_ccitt_false_update() for updating */
    return crc16_ccitt_false_update(buf, len);
}
```

File: sys/include/checksum/crc16_ccitt.h [hunk 5]
```diff
@@ -37,8 +34,92 @@
...
+ * @return          Checksum of the specified memory area
+ */
+uint16_t crc16_ccitt_mcrf4xx_calc(const unsigned char *buf, size_t len);
+
+/**
+ * @brief           Update CRC16-CCITT-FALSE
+ *
+ * @param[in]  crc  A start value for the CRC calculation, usually the
+ *                  return value of a previous call to
+ *                  crc16_ccitt_aug_calc(), crc16_ccitt_false_calc() or crc16_ccitt_false_update()
```
Comment (miri64) [inline sys/include/checksum/crc16_ccitt.h:82]: Of course this needs to be adopted then:
```suggestion
 *                  crc16_ccitt_false_calc() or crc16_ccitt_false_update()
```

File: sys/include/checksum/crc16_ccitt.h [hunk 6]
```diff
@@ -37,8 +34,129 @@
...
+    /* Since CCITT-AUG and CCITT-FALSE only differ in the starting
+     * seed, we wrap around crc16_ccitt_false_update() for updating */
+    return crc16_ccitt_false_update(crc, buf, len);
+}
+
+/**
+ * @brief           Calculate CRC16-CCITT-AUG
+ * @details         polynom=0x1021
+ *                  seed=0x1d0f
+ *                  check=0xe5cc
```
Comment (miri64) [inline sys/include/checksum/crc16_ccitt.h:147]: Is this what you intend the documentation to look like?

Flat

In case you do not and my code-based proposal did not work for you (which would look like this)...

Code

... a table could also be an option:

```suggestion
 *
 * Parameter | Value
 * --------: | :----
 * Polynom   | `0x1021`
 * Seed      | `0x1d0f`
 * Check     | `0xe5cc`
```

Table

Comment (Teufelchen1) [reply sys/include/checksum/crc16_ccitt.h]: > [..] and my code-based proposal did not work for you (which would look like this)...

You proposal was great, I just forgot about it - should not have pushed last minute before heading home... 
I like the table idea - I picked it up and added some other values as well. Without going to much into detail: These parameter specify / identify one particular CRC algorithm, no more naming confusion! I initially omitted them since the `check` is already enough but the tables were just asking for more data. 😀

Thanks for providing your insights, I valued it a lot! :)

**Comment by miri64** [comment]
Please fix the issues pointed out by the [static-tests](https://github.com/RIOT-OS/RIOT/actions/runs/2956539108). Other than that, I think this is ready for ACK now.

File: tests/unittests/tests-checksum/tests-checksum-crc16-ccitt-false.c [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+static void test_checksum_crc16_ccitt_false_sequence_256a(void)
+{
+    unsigned char buf[] = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+                          "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+                          "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+                          "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+                          "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+                          "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
+                          "AAAA";
+    uint16_t expect = 0xEA0B;
```
Comment (miri64) [inline tests/unittests/tests-checksum/tests-checksum-crc16-ccitt-false.c:64]: Do you have a source for the test vectors? If yes, would be great to cite it here in a comment.

**Review by miri64 — APPROVED** [review]
Ran the new tests on `native` and `iotlab-m3`. Found a minor nit. Feel free to address or not.

File: tests/unittests/tests-checksum/tests-checksum-crc16-ccitt-false.c (hunk 7 above)
Comment (Teufelchen1) [reply tests/unittests/tests-checksum/tests-checksum-crc16-ccitt-false.c:64]: I run the test vectors through multiple different implementations of theses crc and checked they matched. E.g. The linux kernel produces the same results for the given vectors. Not sure if it is worth noting in the comments - at the end, it is math. More important it is imo test the API behaves correctly for the edge case empty data.

**Comment by MrKevinWeiss** [comment]
Kicking this out of the CI queue until master is fixed (or blacklisted)

**Comment by MrKevinWeiss** [comment]
Rerunning now that master should pass...
